src/components/ingestion/AbstractIngestionEngine.js

```diff
@@ -1,66 +1,120 @@
 			throw new AbstractInstantiationError(this.constructor.name)
 		}
 		this.producer = kafka.producer()
+		this.mpegtsDemuxer = new TSDemuxer(this.processDemuxedPacket)
+		this.mostRecentDemuxedPacket = null
 	}
 
 	/**
-	 * Ingests data into the TV Kitchen pipeline.
+	 * Ingests an mpegts payload into the TV Kitchen pipeline.
 	 *
-	 * @param  {Buffer} data The buffer of data to be ingested.
-	 * @return {Boolean}     Whether or not the buffer was successfully ingested.
+	 * @param  {Payload} data The payload to be ingested.
+	 * @return {Boolean}      Whether or not the payload was successfully ingested.
 	 */
-	ingestData = async (data) => {
-		const payload = new Payload()
-		payload.data = data
-		await this.producer.send({
-			topic: dataTypes.STREAM.DATA,
-			messages: [{
-				value: data,
-			}],
-		})
-		return true
+	ingestPayloadStream = Writable({
+		objectMode: true,
+		write: async (payload, enc, done) => {
+			if (!(payload instanceof Payload)) {
+				done(new Error('ingestPayloadStream received non-Payload data'))
+			} else {
+				done(this.producer.send({
+					topic: dataTypes.STREAM.CONTAINER,
+					messages: [{
+						value: Payload.serialize(payload),
+					}],
+				}))
+			}
+		},
+	})
+
+	/**
+	 * Updates ingestion state based on the latest demuxed packet data.
+	 *
+	 * This method is called by our mpegts demuxer, and allows the ingestion engine to track
+	 * the most recent demuxed packet.
+	 *
+	 * @param  {Packet} packet The latest TSDemuxer Packet object.
+	 */
+	processDemuxedPacket = (packet) => {
+		this.mostRecentDemuxedPacket = packet
 	}
 
 	/**
-	 * Continuously ingest of a stream of data.
+	 * Returns the most recent cohearant stream packet processed by the ingestion engine.
+	 * This packet is lower level than the mpegts container, and represents an audio or video
+	 * packet demuxed from the mpegts stream.
+	 *
+	 * @return {Packet} The most recent Packet object extracted by TSDemuxer.
+	 */
+	getMostRecentDemuxedPacket = () => this.mostRecentDemuxedPacket
+
+	/**
+	 * Process a new chunk of data from an mpegts stream. The chunks passed to this
+	 * function should be presented sequentially, and combine to form a cohearant mpegts
+	 * data stream, but they can be of arbitray size.
 	 *
-	 * @param  {ReadableStream} readableStream The stream of data to be recorded.
-	 * @return {ReadableStream}                The input stream (for chaining)
+	 * This data is used to update local state and ultimately pipe to kafka.
+	 *
+	 * @param  {Buffer} mpegtsData The latest sequential chunk of mpegts data.
 	 */
-	recordStream = (readableStream) => readableStream
-		.on('data', async (data) => {
-			readableStream.pause()
-			await this.ingestData(data)
-			readableStream.resume()
-		})
+	processMpegtsStream = Transform({
+		objectMode: true,
+		write: (mpegtsData, enc, done) => {
+			this.mpegtsDemuxer.process(mpegtsData)
+			const demuxedPacket = this.getMostRecentDemuxedPacket() || { pts: 0 }
```

Comment:
Is it worthwhile to move `{ pts: 0 }` to a named constant atop the file like `EMPTY_DEMUXED_PACKET` or maybe `this.generateEmptyDemuxedPacket()`? So we can modify its composition from a more visible place?

Reply:
Great thought -- I think this.generateEmptyDemuxedPacket would be good... or maybe it would be a good addition to TSDemuxer specifically.

Reply:
Update: I can just `import { Packet } from 'ts-demuxer'` and then `new Packet()`

Reply:
Hey what about just doing `import { Packet } from 'ts-demuxer'` and then `new Packet()`?